cmake/weak_symbol.cpp

```diff
@@ -0,0 +1,3 @@
+void __attribute((weak)) f();
```

Comment:
Is this the way to write it for all compilers? If not, I can devise a simple CMake loop to determine which, if any, weak symbol attribute notation works.

Reply:
GCC, Clang, [ICC](https://software.intel.com/en-us/articles/performance-tools-for-software-developers-gnu-compatibility-attribute-weak-and-internal-linkage), and XL C [http://www-01.ibm.com/support/docview.wss?uid=swg27044475&aid=7] all use the same weak symbol syntax.

MSVC has something different, but it's not quite the same as weak symbols and would have to be handled differently. Probably best to deal with that whenever we add other pieces for MSVC.

Reply:
OK makes sense to me. Just like to leverage build tools for compatibility when it is straightforward to do so.